README.md
clj-cbor
========

[![CircleCI](https://circleci.com/gh/greglook/clj-cbor.svg?style=shield&circle-token=21efcbc50fe431aa2efc22413ba1f4407fec6283)](https://circleci.com/gh/greglook/clj-cbor)
[![codecov](https://codecov.io/gh/greglook/clj-cbor/branch/main/graph/badge.svg)](https://codecov.io/gh/greglook/clj-cbor)
[![cljdoc](https://cljdoc.org/badge/mvxcvi/clj-cbor)](https://cljdoc.org/d/mvxcvi/clj-cbor/CURRENT)

This library is a native Clojure implementation of the [Concise Binary Object Representation](http://cbor.io/)
format specified in [RFC 7049](https://tools.ietf.org/html/rfc7049).

CBOR is a binary encoding with the goal of small code size, compact messages,
and extensibility without the need for version negotiation. This makes it a good
alternative to [EDN](https://github.com/edn-format/edn) for storing and
transmitting Clojure data in a more compact form.


## Installation

Library releases are published on Clojars. To use the latest version,
add the following dependency to your project definition:

### Leiningen

Add `[mvxcvi/clj-cbor "1.1.0"]` to your `project.clj` file.

### tools.deps

Add `mvxcvi/clj-cbor {:mvn/version "1.1.0"}` to your `deps.edn` file.

## Usage

The `clj-cbor.core` namespace contains the high-level encoding and decoding
functions. The simplest way to use this library is to require it and call them
directly with data:

```clojure
=> (require '[clj-cbor.core :as cbor])

=> (cbor/encode [0 :foo/bar true {:x 'y} #{1/3} #"foo"])
; 0x8600D827683A666F6F2F626172F5A1D827623A78D8276179CD81D81E820103D82363666F6F

=> (cbor/decode *1)
[0 :foo/bar true {:x y} #{1/3} #"foo"]
```

With no extra arguments, `encode` and `decode` will make use of the
`default-codec`, which comes loaded with read and write handler support for many
Java and Clojure types (see the [type extensions](#type-extensions) section
below). Both functions accept an additional argument to specify the codec,
should different behavior be desired.

```clojure
=> (def codec (cbor/cbor-codec :canonical true))

=> (cbor/encode codec {:foo "bar", :baz 123})
; 0xA2D827643A666F6F63626172D827643A62617A187B

=> (cbor/decode codec *1)
{:foo "bar", :baz 123}
```

So far we haven't specified any outputs when encoding, so we've gotten a byte
array back. The full form of `encode` takes three arguments: the codec, the
output stream, and the value to encode.

```clojure
=> (def out (java.io.ByteArrayOutputStream.))

=> (cbor/encode codec out :a)
5

=> (cbor/encode codec out 123)
2

=> (cbor/encode codec out true)
1

=> (cbor/encode codec out "foo")
4

=> (.toByteArray out))
; 0xD827623A61187BF563666F6F

=> (with-open [input (java.io.ByteArrayInputStream. *1)]
     (doall (cbor/decode-seq codec input)))
(:a 123 true "foo")
```

In this mode, `encode` returns the number of bytes written instead of a byte
array. We can read multiple items from an input stream using `decode-seq`, which
returns a lazy sequence. If the input is a file you must realize the values
before closing the input. Similarly, `encode-seq` will write a sequence of
multiple values to an output stream.

As a convenience, the library also provides the `spit`, `spit-all`, `slurp`, and
`slurp-all` functions, which operate on files:

```clojure
=> (cbor/spit "data.cbor" {:abc 123, :foo "qux", :bar true})
29

=> (cbor/spit-all "data.cbor" [[0.0 'x] #{-42}] :append true)
12

=> (.length (io/file "data.cbor"))
41

=> (cbor/slurp "data.cbor")
{:abc 123, :bar true, :foo "qux"}

=> (cbor/slurp-all "data.cbor")
({:abc 123, :bar true, :foo "qux"} [0.0 x] #{-42})
```


## Type Extensions

In order to support types of values outside the ones which are a native to CBOR,
the format uses _tagged values_, similar to EDN. In CBOR, the tags are integer
numbers instead of symbols, but the purpose is the same: the tags convey _new
semantics_ about the following value.

The most common example of a need for this kind of type extension is
representing an instant in time. In EDN, this is represented by the `#inst` tag
on an ISO-8601 timestamp string. CBOR offers two tags to represent instants -
tag 0 codes a timestamp string, while tag 1 codes a number in epoch seconds. The
former is more human-friendly, but the latter is more efficient.

New types are implemented by using read and write handlers - functions which map
from typed value to representation and back. Currently, the library comes with
support for the following types:

|   Tag | Representation | Type | Semantics |
|-------|----------------|------|-----------|
|     0 | Text string    | `Date`/`Instant` | Standard date/time string |
|     1 | Number         | `Date`/`Instant` | Epoch-based date/time |
|     2 | Byte string    | `BigInt` | Positive bignum |
|     3 | Byte string    | `BigInt` | Negative bignum |
|     4 | Array(2)       | `BigDecimal` | Decimal fraction |
|    27 | Array(2)       | `TaggedLiteral` | Constructor support for Clojure tagged literal values |
|    30 | Array(2)       | `Ratio` | Rational fractions, represented as numerator and denominator numbers |
|    32 | Text string    | `URI` | Uniform Resource Identifier strings |
|    35 | Text string    | `Pattern` | Regular expression strings |
|    37 | Byte string    | `UUID` | Binary-encoded UUID values |
|    39 | Text string    | `Symbol`/`Keyword` | Identifiers |
|   100 | Integer        | `LocalDate` | Epoch-based local calendar date |
|   258 | Array          | `Set` | Sets of unique entries |
|  1004 | Text string    | `LocalDate` | String-based local calendar date |
| 55799 | Varies         | N/A | Self-describe CBOR |

For further information about registered tag semantics, consult the
[IANA Registry](https://www.iana.org/assignments/cbor-tags/cbor-tags.xhtml).

### Write Handlers

A write handler is a function which takes a typed value and returns an
encodable representation. In most cases, the representation is a CBOR tagged
value. The tag conveys the type semantic and generally the expected form that
the representation takes. Write handlers are selected by a dispatch function,
which defaults to `class`. The `clj-cbor.core/dispatch-superclasses` function
can be used to construct an inheritance-based dispatcher.

In some cases, multiple types will map to the same tag. For example, by default
this library maps both `java.util.Date` and the newer `java.time.Instant` types
to the same representation.

### Read Handlers

A read handler is a function which takes the representation from a tagged value
and returns an appropriately typed value. The choice of function to parse the
values thus determines the 'preferred type' to represent values of that kind.

Continuing the example, the library comes with read handlers for both `Date` and
`Instant` types, allowing the user to choose their preferred time type.


## Performance

As of `0.7.1`, this library is competitive with many other comparable
serialization formats. Some benchmarking results can be found in
[this spreadsheet](https://docs.google.com/spreadsheets/d/142LhWX5aCnOoF6v0T46RASULQDuG7JIckKiCohDPgq8/edit?usp=sharing).

For small and medium data sizes CBOR is more compact than most formats, while at
larger sizes (above 512 bytes) all formats are fairly close in size (within 10%,
generally). Other than Nippy, which was by far the fastest codec, clj-cbor
was one of the fastest encoders and is in the middle of the pack in decoding
times.

To give some concrete performance numbers, here are a few samples from the
dataset:

| Size |    Encode |    Decode |
|------|-----------|-----------|
|    4 |   6.09 µs |   2.31 µs |
|   55 |  20.87 µs |   7.58 µs |
|  173 |  12.64 µs |   5.74 µs |
|  388 |  15.38 µs |  11.60 µs |
|  882 |  31.55 µs |  14.24 µs |
| 1632 |  54.82 µs |  33.52 µs |
| 3127 |  92.14 µs |  64.66 µs |
| 4918 | 104.92 µs |  59.67 µs |
| 7328 | 108.37 µs |  82.16 µs |


## Notes

A few things to keep in mind while using the library:

- Streaming CBOR data can be parsed from input, but there is currently no way to
  generate streaming output data.
- Decoding half-precision (16-bit) floating-point numbers is supported, but the
  values are promoted to single-precision (32-bit) as the JVM does not have
  native support for them. There is currently no support for writing
  half-precision floats except for the special values `0.0`, `+Inf`, `-Inf`, and
  `NaN`, which are always written as two bytes for efficiency.
- CBOR does not have a type for bare characters, so they will be converted to
  single-character strings when written.
- Regular expressions are supported using tag 35, but beware that Java
  `Pattern` objects do not compare equal or have the same hash code for
  otherwise identical regexes. Using them in sets or as map keys is discouraged.


## License

This is free and unencumbered software released into the public domain.
See the UNLICENSE file for more information.


CHANGELOG.md
Change Log
==========

All notable changes to this project will be documented in this file, which
follows the conventions of [keepachangelog.com](http://keepachangelog.com/).
This project adheres to [Semantic Versioning](http://semver.org/).


## [Unreleased]

...


## [1.1.1] - 2022-06-20

### Changed
- Update Clojure to 1.11.1

### Fixed
- Exclude the new `parse-uuid` function in Clojure 1.11 to avoid a reference
  collision warning.


## [1.1.0] - 2020-06-14

### Added
- The base logic in the codec will write any collection supporting the standard
  Java `List`, `Map`, and `Set` interfaces as the corresponding CBOR types.
  Previously, this only worked directly on Clojure's collection types.
- Local calendar dates are supported using the proposed tags 100 (epoch) and
  1004 (string), mapped to the `java.time.LocalDate` type.


## [1.0.0] - 2020-04-20

### Added
- Errors about unhandled types now contain more type information about the
  value that could not be encoded.
  [#15](https://github.com/greglook/clj-cbor/pull/15)
- The `clj-cbor.core/dispatch-superclasses` function can be used to support
  inheritance-based write handler selection.
  [#12](https://github.com/greglook/clj-cbor/issues/12)


## [0.7.2] - 2019-06-28

### Added
- Add [clj-async-profiler](https://github.com/clojure-goes-fast/clj-async-profiler)
  support in the benchmarking code to produce flame graphs for finding hotspots.

### Removed
- Codecs no longer support a configurable set tag; it is now fixed at 258
  matching the IANA registry.

### Fixed
- `decode` will not try to coerce inputs which are already input streams
  anymore. This fixes incremental/lazy decoding behavior.
  [#13](https://github.com/greglook/clj-cbor/issues/13)
- `encode` will not try to coerce outputs to an output stream anymore. Now an
  exception is thrown when a bad argument is given. This prevents lost writes
  when the final buffer is not flushed correctly.
  [#14](https://github.com/greglook/clj-cbor/pull/14)


## [0.7.1] - 2019-01-10

This is a performance-focused release that introduced a number of changes to
speed up decoding performance. The main change is the switch to a _jump table_
for the initial header byte.
[#9](https://github.com/greglook/clj-cbor/issues/9)
[#11](https://github.com/greglook/clj-cbor/pull/11)

For a representative data sample of about ~97 KB, this brings the benchmarked
decoding time from 8.454 ms to 4.089 ms, or about twice as fast!

### Changed
- Upgrade to Clojure 1.10.0.
- Many operations in the codec are type-hinted to use primitive operations where
  possible.
- CBOR arrays and maps are built up using transients for performance.
- Decoding logic now uses a jump table.

### Fixed
- A tagged-value major type with a streaming info code now results in a
  `::codec/illegal-stream` error.


## [0.6.0] - 2017-12-23

### Changed
- Upgrade to Clojure 1.9.0.


## [0.5.0] - 2017-11-08

This release fixes two of the longer-standing quirks with the library, which
were unfortunately breaking changes. The fixes should be straightforward:

- Replace any `(cbor/decode ...)` with `(cbor/decode-seq ...)`.
- Replace any `(first (cbor/decode-seq ...))` with `(cbor/decode ...)`.

If you have existing encoded data containing sets, you can use the following
function to rewrite it:

```clojure
(defn rewrite-cbor-sets
  [codec source dest]
  (with-open [input (io/input-stream source)
              output (io/output-stream dest)]
    (->>
      input
      (cbor/decode-seq (assoc codec :set-tag 13))
      (cbor/encode-seq codec output))))
```

If rewriting isn't an option, you can support reading sets via tag 13 by
using a custom read handler:

```clojure
(def compatible-codec
  (assoc-in cbor/default-codec [:read-handlers 13] set))
```

### Changed
- **Breaking:** the default set tag is now 258, matching the IANA registry.
  [#6](//github.com/greglook/clj-cbor/issues/6)
- **Breaking:** `clj-cbor.core/decode` now only decodes a single value; previous
  behavior moved to `decode-seq`.
  [#7](//github.com/greglook/clj-cbor/issues/7)

### Added
- `clj-cbor.core/encode-seq` writes a sequence of values to a byte array or
  output stream.
- `clj-cbor.core/spit-all` similarly writes a sequence of values to an output
  file like repeated calls to `spit` with `:append true`.


## [0.4.1] - 2017-05-17

### Fixed
- Resolved an overflow issue when encoding `Long/MIN_VALUE`.
- Integer values are always decoded as `Long` values, no matter their encoded
  size. Previously, numbers between `Short/MAX_VALUE` and `Integer/MAX_VALUE`
  would return `Integer` values.
  [#4](https://github.com/greglook/clj-cbor/issues/4)
  [#5](https://github.com/greglook/clj-cbor/pull/5)


## [0.4.0] - 2017-03-14

### Added
- Implemented canonical mode sorting of map keys and set entries.


## [0.3.0] - 2016-01-05

### Added
- Support self-describe CBOR tag 55799. This provides a 'magic' three-byte
  sequence to simplify detection of the CBOR format. The
  `clj-cbor.core/self-describe` function will wrap the given value with this
  tag.
- Tag codes have been factored out into integer constants where appropriate to
  improve consistency.
- Add `spit`, `slurp`, and `slurp-all` utility functions to the core namespace.

### Changed
- Read handler functions are no longer called with the tag. This greatly
  simplifies their implementation and allows for reuse of existing
  transformation functions as-is.
- Error type `:clj-cbor.codec/illegal-chunk` renamed to `illegal-chunk-type`.

### Fixed
- The `decode` function now properly raises an error when the input ends
  mid-value rather than at a top-level value boundary.


## [0.2.0] - 2016-12-28

This release includes 100% test coverage!

### Added
- UUIDs are supported in binary form using tag 37.
- CBOR error keywords are organized into a hierarchy underneath
  `:clj-cbor.error/encoding-error` and `:clj-cbor.error/decoding-error`.

### Changed
- `clj-cbor.data.model` renamed to `clj-cbor.data.core`.

src/clj_cbor/header.clj
(ns clj-cbor.header
  "Functions for reading and writing CBOR headers."
  (:require
    [clj-cbor.error :as error])
  (:import
    clojure.lang.BigInt
    (java.io
      DataInputStream
      DataOutputStream)))


(def major-types
  "Vector of major type keywords, indexed by the three-bit values 0-7. (§2.1)"
  [:unsigned-integer
   :negative-integer
   :byte-string
   :text-string
   :data-array
   :data-map
   :tagged-value
   :simple-value])


(def ^:private major-type-codes
  "Map of major type keywords to code values."
  (zipmap major-types (range)))


;; ## Encoding Functions

(defn write-leader
  "Writes a header byte for the given major-type and additional info numbers."
  [^DataOutputStream out mtype info]
  (let [header (-> (bit-and (major-type-codes mtype) 0x07)
                   (bit-shift-left 5)
                   (bit-or (bit-and (long info) 0x1F)))]
    (.writeByte out header)))


(defn write-byte
  "Write an unsigned byte (8-bit) value to the data output stream."
  [^DataOutputStream out i]
  (.writeByte out i))


(defn write-short
  "Write an unsigned short (16-bit) value to the data output stream."
  [^DataOutputStream out i]
  (.writeShort out i))


(defn write-int
  "Write an unsigned int (32-bit) value to the data output stream. Coerces the
  value into a signed representation before writing if necessary."
  [^DataOutputStream out i]
  (.writeInt
    out
    (if (<= i Integer/MAX_VALUE)
      i
      (+ Integer/MIN_VALUE (- (dec i) Integer/MAX_VALUE)))))


(defn write-long
  "Write a long (32-bit) value to the data output stream. Coerces the value
  into a signed representation before writing if necessary."
  [^DataOutputStream out i]
  (.writeLong
    out
    (if (<= i Long/MAX_VALUE)
      i
      (+ Long/MIN_VALUE (- (dec i) Long/MAX_VALUE)))))


(defn write
  "Writes a header byte for the given major-type, plus extra bytes to encode
  the given integer code. Always writes the smallest possible representation.
  Returns the number of bytes written."
  ^long
  [^DataOutputStream out mtype i]
  (cond
    (neg? i)
    (error/*handler*
      ::negative-info-code
      (str "Cannot write negative integer code: " i)
      {:code i})

    (<= i 23)
    (do (write-leader out mtype i)
        1)

    (<= i 0xFF)
    (do (write-leader out mtype 24)
        (write-byte out i)
        2)

    (<= i 0xFFFF)
    (do (write-leader out mtype 25)
        (write-short out i)
        3)

    (<= i 0xFFFFFFFF)
    (do (write-leader out mtype 26)
        (write-int out i)
        5)

    (<= i (* -2N Long/MIN_VALUE))
    (do (write-leader out mtype 27)
        (write-long out i)
        9)

    :else
    (error/*handler*
      ::overflow-info-code
      (str "Cannot write integer code requiring 9 bytes of space: " i)
      {:code i})))


;; ## Decoding Functions

(defn decode
  "Determines the major type keyword and additional information encoded by the
  header byte. §2.1"
  [header]
  [(-> header
       (bit-and 0xE0)
       (bit-shift-right 5)
       (bit-and 0x07)
       (major-types))
   (bit-and header 0x1F)])


(def ^:private two-64
  "Constant holding `2^64` for integer manipulation."
  (.shiftLeft BigInteger/ONE 64))


(defn read-byte
  "Read an unsigned byte (8-bit) value from the data input stream. Promotes the
  value to a long for consistency."
  [^DataInputStream in]
  (long (.readUnsignedByte in)))


(defn read-short
  "Read an unsigned short (16-bit) value from the data input stream. Promotes
  the value to a long for consistency."
  [^DataInputStream in]
  (long (.readUnsignedShort in)))


(defn read-int
  "Read an unsigned int (32-bit) value from the data input stream. Promotes the
  value to a long for consistency."
  [^DataInputStream in]
  (bit-and (long (.readInt in)) 0xFFFFFFFF))


(defn read-long
  "Read an unsigned long (64-bit) value from the data input stream. Handles
  overflowing values by promoting them to a bigint.

  https://tools.ietf.org/html/rfc7049#section-1.2"
  [^DataInputStream in]
  (let [i (.readLong in)]
    (if (neg? i)
      (-> (BigInteger/valueOf i)
          (.add two-64)
          (BigInt/fromBigInteger))
      i)))


(defn read-code
  "Reads a size value from the initial bytes of the input stream. Returns
  either a number, the keyword `:indefinite`, or calls the error handler on
  reserved info codes."
  [^DataInputStream in ^long info]
  (if (< info 24)
    ;; Info codes less than 24 directly represent the number.
    info
    ;; Otherwise, signify the number of bytes following.
    (case info
      24 (read-byte in)
      25 (read-short in)
      26 (read-int in)
      27 (read-long in)
      (28 29 30) (error/*handler*
                   ::reserved-info-code
                   (format "Additional information int code %d is reserved."
                           info)
                   {:info info})
      31 :indefinite)))
